core/src/main/java/io/undertow/server/HttpServerExchange.java

```diff
@@ -1716,6 +1716,13 @@ private void closeAndFlushResponse() {
                     getResponseHeaders().put(Headers.CONTENT_LENGTH, "0");
                 }
                 getResponseChannel();
+            } else if (anyAreClear(state, FLAG_RESPONSE_TERMINATED) && !responseChannel.isOpen()) {
+                // UNDERTOW-1664: Http/2 response channels may be closed prior to the connection. There's
+                // no reason to attempt to flush a response for a closed channel but we must ensure
+                // the listeners have been invoked.
+                invokeExchangeCompleteListeners();
+                IoUtils.safeClose(connection);
+                return;
```

Comment:
@fl4via I'd appreciate feedback on this PR when you have time. This block doesn't feel like the best solution, but I was not able to find the call site which closes the response channel leaving an exchange in an incomplete state. If this block is removed from the PR the server logs a lot of errors complaining that the connection was closed uncleanly, so I opted to include handling for the known case so the logging is available for remaining unexpected cases.
Regardless of this block, I think it's critical to invoke completion listeners, otherwise request logging will not occur and resources may not be released (RequestBufferingHandler request buffers, for example).

Reply:
Hi @carterkozak - I believe this https://github.com/undertow-io/undertow/pull/867 fixed this problem already.
Could you please test the problematic scenario you are facing with latest Undertow 2.0.30.Final release 
and confirm the problem was solved?
Just FYI ^^^ @fl4via

Reply:
Thanks for the ping! I thought this could be related as well, but the issue is still reproducible with this commit (I've re-validated using 2.0.30.Final as well).

You can try it yourself using this reproducer with the `Client` and `Server` classes in the `two` package :-) It reproduces with `one` as well, but that's more likely to hit other bugs, and I don't want to conflate the two.

https://github.com/carterkozak/http2-reproducer-undertow

Reply:
Note that in my scenario the Http2ServerConnection is still considered open despite the stream being closed.